embeddings/task/lightning_task/lightning_task.py

```diff
@@ -0,0 +1,170 @@
+        self.test_metrics = metrics.clone(prefix="test/")
+
+    def training_epoch_end(self, outputs: List[Any]) -> None:
+        self._aggregate_and_log_metrics(self.train_metrics)
+
+    def validation_epoch_end(self, outputs: List[Any]) -> None:
+        self._aggregate_and_log_metrics(self.val_metrics, prog_bar=True)
+
+    def test_epoch_end(self, outputs: List[Any]) -> None:
+        self._aggregate_and_log_metrics(self.test_metrics)
+
+    def _aggregate_and_log_metrics(
+        self, metrics: MetricCollection, prog_bar: bool = False
+    ) -> Dict[str, float]:
+        metric_values = metrics.compute()
+        metrics.reset()
+        self.log_dict(metric_values, prog_bar=prog_bar)
+        return metric_values
+
+    def setup(self, stage: Optional[str] = None) -> None:
+        if stage == "fit" and self.hparams.use_scheduler:
+            assert self.trainer is not None
+            train_loader = self.trainer.datamodule.train_dataloader()
+            tb_size = self.hparams.train_batch_size * max(1, self.trainer.gpus)
+            ab_size = self.trainer.accumulate_grad_batches * float(self.trainer.max_epochs)
+            self.total_steps: float = (len(train_loader.dataset) // tb_size) // ab_size
+
+    def configure_optimizers(self) -> Tuple[List[Optimizer], List[Any]]:
+        """Prepare optimizer and schedule (linear warmup and decay)"""
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [
+                    p for n, p in self.named_parameters() if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.hparams.weight_decay,
+            },
+            {
+                "params": [
+                    p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer = AdamW(
+            optimizer_grouped_parameters,
+            lr=self.hparams.learning_rate,
+            eps=self.hparams.adam_epsilon,
+        )
+
+        if self.hparams.use_scheduler:
+            lr_scheduler = self.configure_scheduler(optimizer=optimizer)
+        else:
+            lr_scheduler = []
+
+        return [optimizer], lr_scheduler
+
+    def configure_scheduler(self, optimizer: Optimizer) -> List[Dict[str, Any]]:
+        scheduler = get_linear_schedule_with_warmup(
+            optimizer,
+            num_warmup_steps=self.hparams.warmup_steps,
+            num_training_steps=self.total_steps,
+        )
+        self.lr_scheduler = {"scheduler": scheduler, "interval": "step", "frequency": 1}
+        return [self.lr_scheduler]
+
+
+class HuggingFaceLightningTask(LightningTask[AutoModel], abc.ABC):
+    def __init__(
+        self,
+        num_labels: int,
+        model_name_or_path: str,
+        downstream_model_type: Type["AutoModel"],
+        unfreeze_transformer_from_layer: Optional[int] = None,
```

Comment:
it is done this way because in flair tasks we have the model frozen on the default and then we have the option to fine-tune it; it is also safer for the regular user who may not have enough resources to train whole model

Reply:
In the Transformers library, the whole model is unfrozen. I think it is intuitive, and we should follow this convention. But it is an open question. cc: @laugustyniak @ktagowski